### ISSUE TITLE
Determine formula for broker resource configuration given expected load

### ISSUE THREAD

**Issue by yolocs** [google/knative-gcp#1553]
In addition to a default setup (https://github.com/google/knative-gcp/issues/1552), we want to enable users (with the need) to configure the resource request/limit and hpa. That requires us provide guidance to answer questions like "with this much load, what broker setup I should use and how much footprint it would cause"?

**Comment by yolocs** [comment]
The result would also help: https://github.com/google/knative-gcp/issues/1553

**Comment by grantr** [comment]
This includes testing multiple scenarios, and maybe multiple brokers.

**Comment by grantr** [comment]
This is blocked by #1552. Before we can develop and document the formula for achieving arbitrary load, we should determine what's needed to achieve a default load.

**Comment by Ectelion** [comment]
We've completed additional experiments for various workloads and summarized our configurability recommendations. Below is a snapshot of the doc shared internally (some details and links are removed for now). As can be seen, we focus on summarizing the general trends rather than providing some exact numbers/formulas (which may likely vary from one version to another), although some experimental results are provided. The contents will then be transferred to the public repo as part of #1524.

--------------------

# BrokerCell Parameters

A BrokerCell allows configuring the following parameters for every data plane deployment:

*   CPU/Memory [requests and limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits)
    *   They are _cpuRequest, cpuLimit, memoryRequest, memoryLimit_
    *   They adopt [Kubernetes resource units](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-units-in-kubernetes)
    *   The values will be passed to the deployment as Pod resource requirements
*   Autoscaling parameters:
    *   _minReplicas_ and _maxReplicas_ are the same concepts in [HPA](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/)
    *   _avgCPUUtilization_ represents an [autoscaling metric](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#autoscaling-on-multiple-metrics-and-custom-metrics) of the type “Utilization” tracking the CPU
    *   _avgMemoryUsage_ represents an [autoscaling metric](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#autoscaling-on-multiple-metrics-and-custom-metrics) of the type “AverageValue” tracking the memory consumption
    *   The values will be passed to the HPA for the deployment

Use <code>[kubectl describe](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#describe)</code> to find what values a BrokerCell is using. The default BrokerCell’s name is <code>default</code>.


```
kubectl describe brokercell $BROKERCELL_NAME -n cloud-run-events

Output:
...
Components:
    Fanout:
      AvgCPUUtilization:  50
      AvgMemoryUsage:     1000Mi
      CpuRequest:          1500m
      MaxReplicas:         10
      MemoryLimit:         2500Mi
      MemoryRequest:       2500Mi
      MinReplicas:         1
    Ingress:
      AvgCPUUtilization:  95
      AvgMemoryUsage:     1500Mi
      CpuRequest:          2000m
      MaxReplicas:         3
      MemoryRequest:       2000Mi
      MinReplicas:         1
    Retry:
      AvgCPUUtilization:  95
      AvgMemoryUsage:     2000Mi
      CpuRequest:          2000m
      MaxReplicas:         1
      MemoryRequest:       2000Mi
      MinReplicas:         1
…
```


The values above are the default configuration. Use <code>[kubectl edit](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#editing-resources)</code> to change these values.


```
kubectl edit brokercell $BROKERCELL_NAME -n cloud-run-events
```

# Typical Configurability Use Cases

### **The throughput is lower than expected**

Please check if you may have encountered potential bottlenecks:

*   **Have you reached the default [PubSub quotas](https://cloud.google.com/pubsub/quotas), especially for publisher and subscriber throughputs?** If so, try to request more quotas.
*   **Are your event consumers able to process events at your expected rate?** You should expect lower throughput when your consumers are slow.
*   **Is there any ongoing load increase?** For rapid load increases, it would take a few minutes for the Broker deployments to scale out, during which the throughput would not match the load. If you care about the delay, try [increasing the fanout minReplicas](#) upfront.
*   **Are there enough resources in the cluster for the Broker deployments to scale out?** Even if there are enough allocatable resources, high CPU or memory utilization on nodes in the cluster may put horizontal scaling at risk. For example, given a cluster with 25 n1-standard nodes (with 4 cores) and the overall utilization of 75%, we may end up with an allocatable CPU of 25000m overall, but only 1000m available on each node, which may block the scheduling of pods with a CPU request above 1000m. Try to add more nodes to the cluster or [node affinity and anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity)** **to isolate Broker deployments from the rest of your workloads.


### **I want to improve throughput**

If your load is within the default targets, try [eliminating potential throughput bottlenecks](#). Otherwise, try the following:

*   **Increase fanout deployment _minReplicas_** ([edit the BrokerCell](#) `.spec.components.fanout.minReplicas`) if you want to support a higher egress throughput but is constrained by [lacking backlog-aware autoscaling](#).
*   **Increase retry deployment _minReplicas_** ([edit the BrokerCell](#) `.spec.components.retry.minReplicas`) if you want to support an error rate that is significantly higher than what we [recommend](#).
*   **Increase all Broker deployments’ resource requirements** (e.g. [edit the BrokerCell](#) for a higher ingress memory request `.spec.components.ingress.memoryRequest`) along with HPA if you want to generally support a higher load or support more Brokers/Triggers in a single BrokerCell. However, do notice the resource requirements have an upper bound the same as the node resources.

Configuring a BrokerCell to support a higher load is not always straightforward. Please refer to [Further Readings](#) for some known factors in play.

### **Broker deployments’ memory consumption is way below what’s requested**

The default memory request accounts for the worst case for our default load target so it is expected that in most cases the memory consumption is below what’s requested. It also follows the GKE best practices for configuring requests and limits to minimize extraction risks for pods in [overcommitted state](https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-resource-requests-and-limits).

If you don’t expect getting anywhere near the [load supported](#), you may consider reducing the memory requests. Here are some tips on how to choose a proper memory request for Broker deployments (under your specific load):

1. Use the “Memory Usage” or “Memory Consumption” [Stackdriver metric](https://cloud.google.com/monitoring/api/metrics_kubernetes) (resource type = “Kubernetes Container”) to determine the peak and high percentile values for memory consumption over a representative time period
2. Leave enough room for traffic “spikes”
3. Make sure to keep _avgMemoryUsage auto scaling_ parameter below the memory request (at least 50% is recommended) to lower the chance of OOM

### **Broker deployments request too much resources**

You might want a much “economical” broker setup if you only need it for lightweight tests or run it in a small cluster for dev purposes. If the BrokerCell fails to be ready because the resource is not sufficient, try [lowering the resource requirements via BrokerCell](#).

# Further Readings: Common Trends and Factors at Play

There are many factors and exponentially many more combinations of these factors that affect a BrokerCell’s capacity and resource requirements. Here are the known factors we have explored:

*   RPS
*   Event size
*   The number of Brokers/Triggers
*   The average number of Triggers invoked per event
*   How fast consumers respond to events
*   Consumer error rate
*   Cold start


## **RPS and Event Size**

RPS from the publishers and the average event payload size are used to define the inbound delivery rate, so these two have the most straightforward influence on resource consumption and throughput.

The figures in the section below may give an intuition on how the resource consumption varies for different combinations of rps and payload size, assuming fast consumers and a low failure rate -- more on the other cases later. This is more of an illustrative sample, rather than a reference table. So, please don’t read into numbers closely, as these will vary from one version to another and the specifics of an experimental setup, but we can make some rough estimates:

*   Single Ingress + Fanout + Retry replicas may handle the load up to 10-20Mb/s, depending on a particular RPS/payload ratio. This translates to the total CPU request of 2 (Ingress) + 1.5 (Fanout) + 2 (Retry) = 4.5 units;
*   For the inbound rate of 256 Mb/s (256kb payload), it takes around 2-3 Ingress and 10 Fanout replicas to support the load with the default setup. This translates to the total CPU request of 4 (Ingress) + 15 (Fanout)  + 2 (Retry) = 21 CPU units;
*   Fanout scales more actively, as compared to Ingress;
*   CPU consumption is the main driver for autoscaling for all the deployments under the above assumptions;
*   Other aspects are considered more closely in the next sections;

**Illustrative sample for resource consumption on Ingress**

**Peak CPU usage on Ingress (fast consumers)**
![image](https://user-images.githubusercontent.com/1721518/95898084-5d0c7700-0d43-11eb-8527-0237702a3859.png)


**Peak memory usage on Ingress, Mb; fast consumers (important)**
![image](https://user-images.githubusercontent.com/1721518/95898125-6a296600-0d43-11eb-9f0e-4144cee323f5.png)

**Number of replicas reached**
![image](https://user-images.githubusercontent.com/1721518/95898244-9c3ac800-0d43-11eb-8de5-e47a73eb9ffe.png)

**CPU consumption visuals for the table above**
![image](https://user-images.githubusercontent.com/1721518/95898406-d6a46500-0d43-11eb-8df3-a1a2e3ff4d07.png)

**Illustrative sample for resource consumption on Fanout (fast consumers)**

**Peak CPU usage on Fanout, CPU units**
![image](https://user-images.githubusercontent.com/1721518/95898491-fb98d800-0d43-11eb-9fe7-fcedb96113ee.png)

**Peak memory usage on Fanout, Mb; fast consumers (important)**
![image](https://user-images.githubusercontent.com/1721518/95898577-18cda680-0d44-11eb-95ed-834b6808e558.png)

**Number of replicas needed to support the load given a gradual load increase**
![image](https://user-images.githubusercontent.com/1721518/95900340-97c3de80-0d46-11eb-9424-c7013d9ca278.png)
 
**Number of replicas reached in a rapid load increase**
![image](https://user-images.githubusercontent.com/1721518/95898794-706c1200-0d44-11eb-9cc5-ca0c5938a13a.png)

**CPU consumption visuals for the tables above. Rapid load increases at the beginning of each test may cause an overprovisioning for the first few minutes**
![image](https://user-images.githubusercontent.com/1721518/95898864-8da0e080-0d44-11eb-932b-a5ca9eb0a242.png)

## **Average Triggers Invoked per Event**

The average number of consumers subscribed per event (or the number of activated triggers) directly impacts the outbound delivery rate by multiplying the rate of inbound deliveries. In the experimental samples above, we assumed the average of 10 triggers activated per event. As this number increases, expect a rapidly rising memory consumption on Fanout and, potentially, Retry deployment. This may be more or less noticeable depending on the behavior of consumers (more on this in the section below). Note that the average here is considered not per single Broker, but across all the Brokers within a given data plane (BrokerCell).


## **Responsiveness of Consumers**

The speed at which consumers acknowledge events may have a profound effect on resource usage for Fanout and Retry services. As the delays increase, resource usage shifts from CPU-intensive to memory-intensive pattern. The series of experiments below illustrates the difference between very fast (response within milliseconds) and very slow (minutes of delays) consumers.

![image](https://user-images.githubusercontent.com/1721518/95899099-e4a6b580-0d44-11eb-9290-9d2ae15cc72e.png)

In the latter case, the memory consumption had rapidly surged to 24Gb. 

Besides, Fanout and Retry can maintain a limited volume of concurrently open connections. Thus, in some cases, even a relatively small set of consumers may noticeably degrade the overall throughput.

If consumers are consistently responsive and their throughput is high, those seeking better resource utilization may consider reducing the memory request on Fanout deployment. The _AvgMemoryConsumption_ parameter should also be updated accordingly. We suggest to keep it significantly below (~50%) the memory request. In case event delivery failures are consistently rare, this approach can be generalized to Retry deployment. 


## **Delivery Failure Rate**

As expected by design, increased failure rate first and foremost raises pressure on the Retry deployment. If these failures are not instant, but caused by a timeout, this would mean an increased memory pressure on Fanout as well. Both deployments have a long timeout on the delivery attempt (10 minutes). Hence, when there are multiple such failures, the memory pressure may build up quite significantly, leading to the kind of worst case illustrated in the _“Responsiveness of Consumers”_ section. At some point, you may need to opt for higher memory limits and maxReplicas values for these deployments.

There is an exponential back-off policy for Retry and the consumers that keep failing will eventually be dropped out. However, many such failures happening simultaneously may increase out-of-memory risks.  Especially, the memory consumption for Retry deployment is very sensitive to the absolute number of failing consumers. For example, 100 consumers failing at 10%  error rate would need much more memory than 10 consumers failing at 100% error rate.  


## **Cold Start and Volatile Load**

Rapid swings in the load may have a significant effect on resource consumption. There tend to be CPU and memory surges on rapid load increases, which then typically smooth out. Because the autoscaling policy is based on raw resource consumption metrics, in some cases, one may observe over-provisioning of replicas for a few minutes.

This can be seen from the illustrative samples presented earlier, where there is a considerable difference in the number of provisioned replicas between gradual and rapid load increases. For convenience, the tables above are merged into a single one:
![image](https://user-images.githubusercontent.com/1721518/95899165-0011c080-0d45-11eb-882f-d94f75b79f8d.png)

Related to this is the cold start effect: when a low load is followed by a sudden increase, it may take a few minutes for the deployments to scale out and increase the throughput. 


Measures to mitigate the cold start effect include:

1. Increasing minReplicas parameter -- this is by far the simplest option, but may result in underutilization of resources under low load;
2. Consider lowering the _AvgCPUUtilization_ parameter -- this may speed up the scale out time, but may increase the overprovisioning effect under volatile load;
3. Re-balancing the Fanout configurations for higher resource requests per pod -- this may result in a better trade-off, but can be hard to balance, as this also involves tuning autoscaling parameters;

Note that there is a trade-off between the overprovisioning effect and the cold start effect. Reducing the former one would likely result in amplifying the latter one, and vice versa.
